### DISCUSSION TITLE
Envelope with the sequencer

### DISCUSSION THREAD

**Discussion by bthj** [chaosprint/glicol#109] [Q&A] [answered]
How can an envelope (`envperc`) be used with the sequencer?

In the documentation I see an [example](https://glicol.org/tour#envelope) using it with `imp` and I've used something like:
```
~o3c: imp 0.0625 >> sp \piano5 >> mul ~amp3c >> mul 0.7
~amp3c: imp 0.0625 >> envperc 0.5, 15
```

But I'm unsure how to apply `envperc` to each `seq` note in something like:
```
out: speed 0.2 >> seq 60 _ >> sp \sample 
```

A playable example of where I'd like to try this can be heard at 
https://synth.is/live-coding/01G8EYPK8QYMAV7PWCF1BY6C16
where the samples start with a click, which I'd like to relieve with an envelope (the buttons Apply or Play (for recording steps) can be pressed when all samples have rendered).

**Comment by chaosprint** [comment, marked as answer]
Perhaps you can see this demo:
https://glicol.org/demo#kickdrumsynthesis

The behaviour of the `envperc` is that it will retrigger the envelope everytime it gets a non-zero value from the input. `seq` will output number patterns based on your command. For example, `seq 60` will output one `1.0` in one bar of audio samples. 

```
1.0, 0.0, 0.0, ...
```

You can also use the same `seq` to control pitches as well, which is shown in the example above.

One issue here is that with the lower pitch, the env amp will also be smaller. A temporary solution is to use a `meta` node in between.

```
out: sin ~pitch >> mul ~envb >> mul 0.9

~envb: ~triggerb >> meta `
    output = input.map(|i|
		if i > 0.0 {
			1.0
		} else {
			0.0
		}
	);
    output
` >> envperc 0.01 0.4;

~env_pitch: ~triggerb >> envperc 0.01 0.1;

~pitch: ~env_pitch >> mul 50 >> add 60;

~triggerb: speed 4.0 >> seq 90
```

**Reply by bthj** [reply]
Thanks @chaosprint - for one sample I now have the desired effect with something like:

```
out: ~triggerb >> sp \mySample >> mul ~envb >> mul 0.4
~envb: ~triggerb >> envperc 0.1 4.4
~triggerb: speed 0.4 >> seq 60
```

But I wonder if it would make sense to support syntax like the following for the same effect:
```
out: speed 0.4 >> seq 60 >> sp \mySample >> envperc 0.1 4.4 >> mul 0.4
```
?

**Reply by bthj** [reply]
...or can I somehow wrap the 
```
~envb: ~triggerb >> envperc 0.1 4.4
~triggerb: speed 0.4 >> seq 60
```
part in a reusable function for different sequencing patterns?

**Reply by bthj** [reply]
...am thinking about how to reduce code repetition in something like:
```
~n1: ~triggerb_n1 >> sp \mySample1 >> mul ~envb_n1 >> mul 0.4
~envb_n1: ~triggerb_n1 >> envperc 0.1 5.5
~triggerb_n1: speed 0.25 >> seq 60 _

~n2: ~triggerb_n2 >> sp \mySample2 >> mul ~envb_n2 >> mul 0.4
~envb_n2: ~triggerb_n2 >> envperc 0.1 5.5
~triggerb_n2: speed 0.25 >> seq _ 60

out: mix ~n.. >> plate 0.1
```
where the main difference is the seq pattern (`60 _` vs `_ 60`) - the envelope is otherwise the same.

**Reply by chaosprint** [reply]
> 

Interesting question. So far I can only reduce it to this:
```
~trigger_n1: speed 0.5 >> seq 60 _
~envb_n1: ~trigger_n1 >> envperc 0.01 0.1
~n1: ~trigger_n1 >> sp \808clap >> mul ~envb_n1

~trigger_n2: speed 0.5 >> seq _ 60 
~n2: ~trigger_n2 >> sp \blip >> mul ~envb_n2
~envb_n2: ~trigger_n2 >> envperc 0.01 0.1

out: mix ~n.. >> mul 0.4 >> plate 0.1
```

**Reply by bthj** [reply]
Thanks @chaosprint, that's a good alternative.  It's more intuitive to write out the sequencer first like that, as a trigger for the envelope and sample that follow.
